src/osd/PrimaryLogPG.cc

```diff
@@ -3261,12 +3266,23 @@ struct RefCountCallback : public Context {
 };
 
 struct SetManifestFinisher : public PrimaryLogPG::OpFinisher {
+  PrimaryLogPGRef pg;
+  ObjectContextRef rollback_to;
   OSDOp& osd_op;
+  PrimaryLogPG::OpContext *ctx;
 
-  explicit SetManifestFinisher(OSDOp& osd_op) : osd_op(osd_op) {
+  explicit SetManifestFinisher(PrimaryLogPG *p, ObjectContextRef r,
+      OSDOp& osd_op, PrimaryLogPG::OpContext *ctx) : 
+    pg(p), rollback_to(r), osd_op(osd_op), ctx(ctx) {
   }
 
   int execute() override {
+    // if rollback_to is not nullptr, this is a callback from rollback_to()
+    if (rollback_to) {
```

Comment:
This gets invoked in do_osd_ops in the case CEPH_OSD_OP_ROLLBACK block.  Please move this logic and the finisher check into _rollback_to.  That way, the user would see the op_finisher setup and completion in the same place.  It also eliminates this inconvenient asymmetry between this caller and the other two.